compiler/Repl.hs

```diff
@@ -452,19 +468,96 @@ execFile name line = do
-
-replFrom :: DebugMode -> [FilePath] -> IO ()
-replFrom mode files = defaultState mode >>= evalStateT (loadFiles files >> runInputT defaultSettings runRepl)
+outputDoc :: (MonadState ReplState m, MonadIO m) => Doc -> m ()
+outputDoc x = do
+  h <- gets outputHandle
+  liftIO $ do
+    hPutDoc h x
+    hFlush h
+
+runRemoteReplCommand :: Int -> String -> IO ()
+runRemoteReplCommand port command = Net.withSocketsDo $ do
+  sock <- Net.socket Net.AF_INET Net.Stream Net.defaultProtocol
+  r <- try $ Net.connect sock (Net.SockAddrInet (fromIntegral port) $ Net.tupleToHostAddress (127, 0, 0, 1))
+  case r of
+    Right () -> do
+      handle <- Net.socketToHandle sock ReadWriteMode
+      hSetEncoding handle utf8
+      T.hPutStrLn handle (T.pack command)
+      T.putStr =<< T.hGetContents handle
+    Left (_ :: SomeException) ->
+      putStrLn $ "Failed to connect to server on port " ++ show port
+
+repl :: Int -> DebugMode -> IO ()
+repl port mode = replFrom port mode []
+
+replFrom :: Int -> DebugMode -> [FilePath] -> IO ()
+replFrom port mode files = do
+  state <- defaultState mode
+  hSetBuffering stdout LineBuffering
+
+  tid <-
+    if port /= 0
+       then forkIO $ startServer port files state
+       else myThreadId
+
+  evalStateT (loadFiles files >> runInputT defaultSettings (runRepl (Just tid))) state
+
+finish :: MonadIO m => Listener -> m ()
+finish Nothing = pure ()
+finish (Just i) = liftIO $ throwTo i ThreadKilled
+
+startServer :: Int -> [FilePath] -> ReplState -> IO ()
+startServer port files state = Net.withSocketsDo $ do
+  let getSock = do
+        (addr:_) <-
+          Net.getAddrInfo (Just (Net.defaultHints { Net.addrFlags = [ Net.AI_NUMERICHOST
+                                                                    , Net.AI_PASSIVE 
+                                                                    ]
+                                                  , Net.addrSocketType = Net.Stream }))
+                          (Just "127.0.0.1")
+                          (Just (show port))
+        sock <- Net.socket Net.AF_INET Net.Stream Net.defaultProtocol
+        x <- try $ do
+          Net.setSocketOption sock Net.ReuseAddr 1
+          Net.bind sock (Net.addrAddress addr)
+          Net.listen sock 4
+        case x of
+          Left (_ :: SomeException) -> do
+            putStrLn $ "Failed to start REPL server on port " ++ show port
+            killThread =<< myThreadId
+          Right () -> pure ()
+        putStrLn $ "Listening on port " ++ show port
+        pure sock
+
+      handleReplLine line =
+        if | T.null line -> pure ()
+           | ':' == T.head line -> uncurry (execCommand Nothing) . span (/=' ') $ T.unpack (T.tail line)
+           | otherwise -> () <$ execString "=<remote command>" line
+
+
+  nulldev <- openFile "/dev/null" WriteMode
+  bracket getSock Net.close $ \sock -> do
+    state <- execStateT (loadFiles files) (state { outputHandle = nulldev })
```

Comment:
OK, this seems crazy dubious.

Reply:
oh yeah it really is but I couldn't think of any better way. we don't want it to print the loading message twice (once for the server, once for the repl itself), so I just did that instead

Reply:
ah quick fix load once before starting either I suppose

Reply:
gimme a second I'm on it